File: lib/src/main/java/anetos/software/lib/SparkSegment.java
package anetos.software.lib;

public class SparkSegment {

    public static String spark(int length) {
        if (length == 0)
            return "";

        String pre = "X";
        for (int i = 0; i < length; i++) {
            pre = pre.concat("~");
        }
        return pre;
    }

    public static String freeSegments(int length) {
        String suffix = "";
        if (length == 0) {
            return "";
        } else {
            for (int i = 0; i < length; i++) {
                suffix = suffix.concat("0");
            }
            return suffix;
        }
    }

    static int traversedPosition = 0;

    public static String recursive(int[] a, int position, int length) {
        String pre = "";
        String preTemp = "";
        while (position <= a.length - 1) {
            if (pre.length() < length) {
                preTemp = pre;
                pre = pre.concat(spark(a[position]));
                if (pre.length() > length) {
                    return preTemp;
                }
            } else {
                break;
            }
            position++;
            traversedPosition = position;
        }

        return pre;
    }

    public static void main(String[] args) {

        int[] array = {0, 9, 9, 8, 3, 5, 6, 3, 2, 2, 2, 5};
        int width = 10;

        for (int value : array) {
            if (value + 1 > width) {
                System.out.println("\uD83D\uDE14   Element you've added is out of range.");
                System.exit(0);
            }
        }

        for (int i = 0; i < array.length; i++) {
            int first = array[i] + 1;
            if (first == width) {
                System.out.println(spark(array[i]));
            } else {
                String output = recursive(array, i, width);
                if (output.length() < width) {
                    System.out.println(output.concat(freeSegments(width - output.length())));
                } else {
                    System.out.println(output);
                }
                i = traversedPosition - 1;
            }
        }
    }
}

File: kotlin_lib/src/main/java/anetos/software/kotlin_lib/SparkSegment.kt
package anetos.software.kotlin_lib

object SparkSegment {
    fun spark(length: Int): String {
        if (length == 0) return ""
        var pre = "X"
        for (i in 0 until length) {
            pre = pre + "~"
        }
        return pre
    }

    fun freeSegments(length: Int): String {
        var suffix = ""
        return if (length == 0) {
            ""
        } else {
            for (i in 0 until length) {
                suffix = suffix + "0"
            }
            suffix
        }
    }

    var traversedPosition = 0
    fun recursive(a: IntArray, position: Int, length: Int): String {
        var position = position
        var pre = ""
        var preTemp = ""
        while (position <= a.size - 1) {
            if (pre.length < length) {
                preTemp = pre
                pre = pre + spark(a[position])
                if (pre.length > length) {
                    return preTemp
                }
            } else {
                break
            }
            position++
            traversedPosition = position
        }
        return pre
    }

    @JvmStatic
    fun main(args: Array<String>) {
        println("KOTLIN")
        val array = intArrayOf(0, 9, 9, 8, 3, 5, 6, 3, 2, 2, 2, 5)
        val width = 10
        for (value in array) {
            if (value + 1 > width) {
                println("\uD83D\uDE14   Element you've added is out of range.")
                System.exit(0)
            }
        }

        var i = 0
        while (i < array.size) {
            val first = array[i] + 1
            if (first == width) {
                println(spark(array[i]))
            } else {
                val output =
                    recursive(array, i, width)
                if (output.length < width) {
                    println(output + freeSegments(width - output.length))
                } else {
                    println(output)
                }
                i = traversedPosition - 1
            }
            i++
        }
    }
}